browser/permissions/localhost_access_permission_browsertest.cc

```diff
@@ -0,0 +1,287 @@
+    current_browser_ = InProcessBrowserTest::browser();
+  }
+
+  void TearDownInProcessBrowserTestFixture() override {
+    mock_cert_verifier_.TearDownInProcessBrowserTestFixture();
+    InProcessBrowserTest::TearDownInProcessBrowserTestFixture();
+  }
+
+  void TearDownOnMainThread() override { prompt_factory_.reset(); }
+
+  permissions::MockPermissionPromptFactory* prompt_factory() {
+    return prompt_factory_.get();
+  }
+
+  Browser* browser() { return current_browser_; }
+
+  void SetBrowser(Browser* browser) { current_browser_ = browser; }
+
+  void SetPromptFactory(permissions::PermissionRequestManager* manager) {
+    prompt_factory_ =
+        std::make_unique<permissions::MockPermissionPromptFactory>(manager);
+  }
+
+  content::WebContents* contents() {
+    return browser()->tab_strip_model()->GetActiveWebContents();
+  }
+
+  void ClickButtonWithId(const std::string& id) {
+    std::string click_script = content::JsReplace(
+        R"(
+        new Promise(async (resolve, reject) => {
+            try {
+              const button = document.getElementById($1);
+              button.click();
+              resolve(true);
+            } catch(error) {
+              reject(error);
+            }
+          })
+      )",
+        id);
+    ASSERT_EQ(true, EvalJs(contents(), click_script));
+  }
+
+  void CheckCurrentStatusIs(ContentSetting content_setting) {
+    EXPECT_EQ(content_settings()->GetContentSetting(
+                  embedding_url_, embedding_url_,
+                  ContentSettingsType::BRAVE_LOCALHOST_ACCESS),
+              content_setting);
+  }
+
+  void SetCurrentStatus(ContentSetting content_setting) {
+    content_settings()->SetContentSettingDefaultScope(
+        embedding_url_, embedding_url_,
+        ContentSettingsType::BRAVE_LOCALHOST_ACCESS, content_setting);
+  }
+
+  void CheckAskAndAcceptFlow(std::string button_id,
+                             GURL localhost_url,
+                             int prompt_count = 0) {
+    CheckCurrentStatusIs(ContentSetting::CONTENT_SETTING_ASK);
+    EXPECT_EQ(prompt_count, prompt_factory()->show_count());
+    ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), embedding_url_));
+    // Accept prompt.
+    prompt_factory()->set_response_type(
+        permissions::PermissionRequestManager::ACCEPT_ALL);
+    ClickButtonWithId(button_id);
+    prompt_factory()->WaitForPermissionBubble();
+    // Make sure prompt came up.
+    EXPECT_EQ(prompt_count + 1, prompt_factory()->show_count());
+    // Check content setting is now ALLOWed.
+    CheckCurrentStatusIs(ContentSetting::CONTENT_SETTING_ALLOW);
+    // Access to localhost resources should be allowed.
+    content::TestNavigationObserver load_observer(contents());
+    ClickButtonWithId(button_id);
+    load_observer.Wait();
+    EXPECT_TRUE(load_observer.last_navigation_succeeded());
+    EXPECT_EQ(localhost_url, load_observer.last_navigation_url());
+  }
+
+  void CheckAskAndDenyFlow(std::string button_id,
```

Comment:
Should we also test the user dismissing the prompt (i.e. not clicking Accept or Deny)?